### PR TITLE
Add RtpTransceiver methods stop, stopped, setDirection, and setCodecPreferences.

### PR DESCRIPTION
**Pull request by pthatcherg**
Fix https://github.com/rtcweb-wg/jsep/issues/261

### CODE REVIEW

**Comment by ekr** [comment]
@pthatcherg: needs rebase

File: draft-ietf-rtcweb-jsep.xml [hunk 1]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
           simulcast as described in 
           <xref target="sec.simulcast" />).</t>
         </section>
+
+        <section title="stopTransceiver" anchor="sec.transceiver-stop">
+          <t>
+            The stopTransceiver method stops an RtpTransceiver.
+            See <xref target="sec.transceiver-stopped"/> for more details.
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: "see below" seems more appropriate when this is the next section

File: draft-ietf-rtcweb-jsep.xml [hunk 2]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+          </t>
+        </section>
+        <section title="isTransceiverStopped" anchor="sec.transceiver-stopped">
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: Did we decide that to do methodFoo() or methodFoo

File: draft-ietf-rtcweb-jsep.xml [hunk 3]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+          <t>
+            The isTransceiverStopped method returns true if
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: "true" and "false" need to be quoted

File: draft-ietf-rtcweb-jsep.xml [hunk 4]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+            the transceiver has been stopped, either by a call
+            to stopTransceiver or by a call to
+            setRemoteDescription, and false otherwise.
+          </t>
+          <t>
+            A stopped RtpTransceiver does not send any
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: Can stopped transceivers be re-started?

File: draft-ietf-rtcweb-jsep.xml [hunk 5]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+            outgoing RTP or RTCP or process any incoming RTP
+            or RTCP.
+          </t>
+        </section>
+        <section title="setTransceiverDirection" anchor="sec.transceiver-set-direction">
+          <t>
+            The setTransceiverDirection method sets the
+            direction of a transceiver, which affects the
+            direction attribute of an associted m=section on
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: associated

File: draft-ietf-rtcweb-jsep.xml [hunk 6]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+            createAnswer.
+          </t>
+          <t>
+            The codec preferences of an RtpTransceiver can
+            affect calls to createOffer and createAnswer by
+            either excluding codecs (in which case the
+            corresponding media formats will be excluded) or
+            by changing the order of codecs (in which case the
+            order of the media formats will match).  The codec
+            preferenes cannot add media formats that would
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: preferences.

Also, can this cause onnegotiationneeded.

File: draft-ietf-rtcweb-jsep.xml (hunk 1 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: Done

**Comment by pthatcherg** [comment]
That's twice now.  Stop moving stuff around!

File: draft-ietf-rtcweb-jsep.xml (hunk 2 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: methodFoo is consistent with the existing text.

File: draft-ietf-rtcweb-jsep.xml (hunk 3 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: True.

For some definition of "true".

File: draft-ietf-rtcweb-jsep.xml (hunk 4 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: no

File: draft-ietf-rtcweb-jsep.xml (hunk 6 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: done

Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: onnegotiationneeded is covered in the W3C spec.  I don't see anywhere in this spec anything that references onnegotiationneeded.   Is there one?

File: draft-ietf-rtcweb-jsep.xml [hunk 7]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
           simulcast as described in
           <xref target="sec.simulcast" />).</t>
         </section>
+
+        <section title="stopTransceiver" anchor="sec.transceiver-stop">
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: To avoid confusion, rather than jamming these on PeerConnection, I think we probably want to make a separate section for RtpTransceivers, e.g.

    4.  Interface
         4.1.  PeerConnection
           4.1.1.  Constructor
           4.1.2.  addTrack
           4.1.3.  addTransceiver
         4.2.  RtpTransceiver
           4.2.1. setDirection 
           4.2.2. stop
           4.2.3. isStopped

(we can probably remove the empty 'Methods' section)
   
...

File: draft-ietf-rtcweb-jsep.xml (hunk 2 above)
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: I think we typically use parens though when referring to a method.

File: draft-ietf-rtcweb-jsep.xml [hunk 8]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+          <t>
+            The isTransceiverStopped method returns true if
+            the transceiver has been stopped, either by a call
+            to stopTransceiver or by a call to
+            setRemoteDescription, and false otherwise.
+          </t>
+          <t>
+            A stopped RtpTransceiver does not send any
+            outgoing RTP or RTCP or process any incoming RTP
+            or RTCP.
```
Comment (fluffy) [inline draft-ietf-rtcweb-jsep.xml]: Might help to add a note that stopped will correspond to a m line with a zero port and that a m-line that has a direction of "inactive" will not be stopped.

File: draft-ietf-rtcweb-jsep.xml [hunk 9]
```diff
@@ -1097,6 +1097,54 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+          <t>
+            The setCodecPreferences method sets the codec
+            preferences of a transceiver, which affect the
+            presence and order of codecs of an associated
+            m=section on future calls to createOffer and
+            createAnswer.
+          </t>
+          <t>
+            The codec preferences of an RtpTransceiver can
+            affect calls to createOffer and createAnswer by
```
Comment (fluffy) [inline draft-ietf-rtcweb-jsep.xml]: I think it can impact createOffer but I don't think it should impact createAnswer. The answer should accept the highest priority codec in the offer regardless of any priorities of this browser or the JS running on side processing the answer.

File: draft-ietf-rtcweb-jsep.xml [hunk 10]
```diff
@@ -2501,6 +2551,10 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <t>The &lt;proto&gt; field MUST be set to exactly match the
             &lt;proto&gt; field for the corresponding m= line in the
             offer.</t>
+
+            <t>If codec preferences have been set for the
+            associated RtpTransceiver, media formats MUST
+            be generated in the corresponding order.</t>
```
Comment (fluffy) [inline draft-ietf-rtcweb-jsep.xml]: I'm not reading this in context but this should only be for Offer not answer. Perhaps it is OK as it.

File: draft-ietf-rtcweb-jsep.xml [hunk 11]
```diff
@@ -2531,7 +2585,8 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <t>A direction attribute which is the same as that of the
             associated transceiver.</t>
 
-            <t>For each supported codec that is present in the offer,
+            <t>For each supported codec that is present in the offer
+            and not excluded by the codec preferences of associated transceiver,
```
Comment (fluffy) [inline draft-ietf-rtcweb-jsep.xml]: again offer only not answer but perhaps this is right

File: draft-ietf-rtcweb-jsep.xml (hunk 8 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: I added a note about making a zero port.  

But what's this about inactive m= sections not being stopped?  That's new to me.

File: draft-ietf-rtcweb-jsep.xml (hunk 9 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: Then how does an answerer specify which codec the prefer to receive?  The W3C spec specifically says "This method allows applications to disable the negotiation of specific codecs. It also allows an application to cause a remote peer to prefer the codec that appears first in the list for sending."  Would that not apply to both the offerer and the answerer?

File: draft-ietf-rtcweb-jsep.xml (hunk 7 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: That's a bit more remodeling of the text than I wanted to do.  Can we land this as-is and then do another PR to remodel?  Or would you like to make a PR of my PR?  I have a feeling you typing what you want will be faster than you typing a description of what you want me to type.

File: draft-ietf-rtcweb-jsep.xml (hunk 2 above)
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: Ever place I see addTrack and addTransceiver don't have parens.  Are we supposed to change them all?

File: draft-ietf-rtcweb-jsep.xml (hunk 9 above)
Comment (fluffy) [inline draft-ietf-rtcweb-jsep.xml]: So 3264 has the text 

   Although the answerer MAY list the formats in their desired order of
   preference, it is RECOMMENDED that unless there is a specific reason,
   the answerer list formats in the same relative order they were
   present in the offer.  In other words, if a stream in the offer lists
   audio codecs 8, 22 and 48, in that order, and the answerer only
   supports codecs 8 and 48, it is RECOMMENDED that, if the answerer has
   no reason to change it, the ordering of codecs in the answer be 8,
   48, and not 48, 8.  This helps assure that the same codec is used in
   both directions.

Which I think supports you argument that if the JS going the create answer set the a preference, then we reorder the codecs in the offer to match that preference when generating the answer

**Comment by pthatcherg** [comment]
Parens added.

File: draft-ietf-rtcweb-jsep.xml (hunk 8 above)
Comment (taylor-b) [inline draft-ietf-rtcweb-jsep.xml]: So for an inactive transceiver, isTransceiverStopped will return true? Then how do you tell the difference between an inactive port 0 transceiver and a plain old inactive one?

File: draft-ietf-rtcweb-jsep.xml (hunk 9 above)
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: Also: what should happen if A offers codecs X and Y but B prefers Y over X, and does a reoffer?

I generally think this should only affect initial offers.

File: draft-ietf-rtcweb-jsep.xml (hunk 4 above)
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: that was supposed to be rhetorical. I think you should change it :)

File: draft-ietf-rtcweb-jsep.xml (hunk 6 above)
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: I think you might say "this may require a new offer/answer transaction"

File: draft-ietf-rtcweb-jsep.xml [hunk 12]
```diff
@@ -1901,8 +1962,9 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
               <t>A direction attribute which is the same as that of the
               associated transceiver.</t>
 
-              <t>For each supported codec, "a=rtpmap" and "a=fmtp"
-              lines, as specified in
+              <t>For each supported codec,
+              not excluded by the codec preferences of associated transceiver,
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: For each supported codec **that has not been** excluded by the codec preferences of **the** associated transceiver

File: draft-ietf-rtcweb-jsep.xml [hunk 13]
```diff
@@ -2304,7 +2366,8 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
 
             <t>The m= line and corresponding "a=rtpmap" and "a=fmtp"
             lines MUST only include codecs present in the most recent
-            answer.</t>
+            answer which are not excluded by the codec preferences of
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: are -> have not been
of -> of the

File: draft-ietf-rtcweb-jsep.xml [hunk 14]
```diff
@@ -2501,6 +2564,10 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <t>The &lt;proto&gt; field MUST be set to exactly match the
             &lt;proto&gt; field for the corresponding m= line in the
             offer.</t>
+
+            <t>If codec preferences have been set for the
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: I think you want to remove this, per our discussion.

File: draft-ietf-rtcweb-jsep.xml [hunk 15]
```diff
@@ -2531,7 +2598,8 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <t>A direction attribute which is the same as that of the
             associated transceiver.</t>
 
-            <t>For each supported codec that is present in the offer,
+            <t>For each supported codec that is present in the offer
+            and not excluded by the codec preferences of associated transceiver,
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: of the

**Comment by taylor-b** [comment]
Alright, I've addressed all the comments so far (except for "transceiver with inactive direction should return true for isStopped", which I disagree with.

Most notably, we now have the rule that the codec preferences can *remove* codecs at any time, but can only *reorder* them for the initial offer. This prevents codecs being reordered back and forth on subsequent offer/answer exchanges, and prevents having different orders in the offer and answer.

File: draft-ietf-rtcweb-jsep.xml [hunk 16]
```diff
@@ -1577,7 +1577,7 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             The isStopped method returns "true" if
             the transceiver has been stopped, either by a call
             to stopTransceiver or by a call to
-            setLocalDescription, and "false" otherwise.
+            setRemoteDescription, and "false" otherwise.
```
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: I think there's a problem here.   An offer can't stop a transceiver, because it's possible that will be rolled back.  An answer must be the thing that can stop a transceiver.   And that would only happen if the remote side offered to stop the transceiver, which means the answer that actually stops it (in a way that can't be rolled back) is the setLocalDescription(answer), not the setRemoteDescription(offer).

File: draft-ietf-rtcweb-jsep.xml [hunk 17]
```diff
@@ -1602,14 +1602,22 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+          <t>
+            The codec preferences of an RtpTransceiver can also determine
+            the order of codecs in subsequent calls to createOffer (if an
+            answer containing the transceiver has not yet been applied),
+            in which case the order of the media formats in the associated
+            m= section will match. The codec preferences cannot change the
+            order of codecs in an answer, and cannot change the order of
+            codecs after an answer containing the transceiver has been
+            applied.
           </t>
```
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: Ah, I see you addressed the "change on re-offer" part of this by restricting it to the initial offer.  But that definitely contradicts the W3C spec, which says "Codec preferences remain in effect for all calls to createOffer and createAnswer that include this RTCRtpTransceiver until this method is called again."

If this is what we really want, then I think we need to go change it in the W3C to match.  And I don't think it's what we decided on in the W3C before.

However, I personally don't care either way.  I think anyone who wants to change what codec to send should just use RtpSender.setParameters.

File: draft-ietf-rtcweb-jsep.xml [hunk 18]
```diff
@@ -1980,6 +1980,10 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
               <xref target="I-D.ietf-rtcweb-video"></xref>(see Section
               5).</t>
 
+              <t>If codec preferences have been set for the associated
+              transceiver, media formats MUST be generated in the
+              corresponding order.</t>
+
```
Comment (pthatcherg) [inline draft-ietf-rtcweb-jsep.xml]: I think it will be a very confusing API if this happens to a developer:

1.  Set the codec preferences.  Receive a call.  Codec preferences ignored.
2.  Set the codec preferences again (or mid-call).  Re-offer.  Codec preferences ignored.
3.  Learn to remove all but my favorite codec.  It works!  Until it doesn't.
4.  Receive a call from an endpoint that doesn't support my favorite codec.  No video.
5.  Learn to check RtpTransceiver.sender.getParameters().codecs first, then pick my favorite out of that group and then exclude the rest.  It works!

Still better than SDP munging?

File: draft-ietf-rtcweb-jsep.xml (hunk 17 above)
Comment (taylor-b) [inline draft-ietf-rtcweb-jsep.xml]: Well they do remain in effect with these rules... just only for the purpose or removing codecs. :)

But I understand your reasoning.

**Comment by taylor-b** [comment]
I looked through our mailing archives and meeting minutes to see what we've previously decided.

When this API was first discussed on the mailing list, [it was proposed](https://lists.w3.org/Archives/Public/public-webrtc/2015Oct/0035.html) that the preferences *do* affect `createAnswer`, and no one voiced any opposition.

Then a couple weeks later [at TPAC 2015](https://www.w3.org/2015/10/28-webrtc-minutes#item04), we decided that the order in `setCodecPreferences` affects the initial offer *and* answer, but not subsequent offers/answers.

So maybe we should just go with what was previously decided?

File: draft-ietf-rtcweb-jsep.xml (hunk 18 above)
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: This is a good argument. You've convinced me we should apply this for initial offers and answers (which sounds like the decision the W3C came to).

**Comment by taylor-b** [comment]
Ready for review again. Rules about codec preferences now match the decision at TPAC 2015 described above.

File: draft-ietf-rtcweb-jsep.xml [hunk 19]
```diff
@@ -1562,6 +1563,74 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+            The codec preferences of an RtpTransceiver can also determine
+            the order of codecs in subsequent calls to createOffer and
+            createAnswer, in which case the order of the media formats in
+            the associated m= section will match. However, the codec
+            preferences cannot change the order of the media formats after
+            an answer containing the transceiver has been applied. At this
+            point, codecs can only be removed, not reordered.
+          </t>
+          <t>
+            Note that, as described in <xref target="RFC3264">
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: This guidance is somewhat irrelevant here since we have a specific set of directives below indicating what should be sent. (i.e., the 'applying a remote description' section). 

So, you could either point at that, or omit this clause and say 'Note that the order of codecs in sCP does not affect which codec the implementation decides to send. It only affects ...'

File: draft-ietf-rtcweb-jsep.xml [hunk 20]
```diff
@@ -1901,15 +1970,20 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+              the codec preferences of the associated transceiver,
+              "a=rtpmap" and "a=fmtp" lines, as specified in
               <xref target="RFC4566"></xref>, Section 6. The audio and
               video codecs that MUST be supported are specified in
               <xref target="I-D.ietf-rtcweb-audio"></xref>(see Section
               3) and
               <xref target="I-D.ietf-rtcweb-video"></xref>(see Section
               5).</t>
 
+              <t>If codec preferences have been set for the associated
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: I think this needs to go into the discussion of the m= line generation itself (e.g. the section after "Each m= section should be generated as specified in [RFC4566],
   Section 5.14.  For the m= line itself, the following rules MUST be
   followed:")

File: draft-ietf-rtcweb-jsep.xml [hunk 21]
```diff
@@ -2541,6 +2620,10 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <xref target="I-D.ietf-rtcweb-video"></xref>(see Section
             5).</t>
 
+            <t>If codec preferences have been set for the associated
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: I think this should go into the m= line generation discussion.

File: draft-ietf-rtcweb-jsep.xml [hunk 22]
```diff
@@ -1562,6 +1563,70 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+            outgoing RTP or RTCP or process any incoming RTP
+            or RTCP.  It cannot be restarted.
+          </t>
+        </section>
+        <section title="setDirection" anchor="sec.transceiver-set-direction">
+          <t>
+            The setTransceiverDirection method sets the
+            direction of a transceiver, which affects the
+            direction attribute of the associated m= section on
+            future calls to createOffer and createAnswer.
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: This seems a bit confusing. It's not like you can unilaterally change something that's recvonly to sendonly without a re-offer. 

How about a TODO for an example.

File: draft-ietf-rtcweb-jsep.xml [hunk 23]
```diff
@@ -1562,6 +1563,70 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+          </t>
+        </section>
+        <section title="setCodecPreferences" anchor="sec.transceiver-set-codec-preferences">
+          <t>
+            The setCodecPreferences method sets the codec
+            preferences of a transceiver, which affect the
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: affects

File: draft-ietf-rtcweb-jsep.xml [hunk 24]
```diff
@@ -1562,6 +1563,70 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
+            preferences cannot change the order of the media formats after
+            an answer containing the transceiver has been applied. At this
+            point, codecs can only be removed, not reordered.
+          </t>
+          <t>
+            Note that the order of codecs in setCodecPreferences does not
+            affect which codec the implementation decides to send. It only
+            affects which codecs the implementation indicates that it prefers
+            to receive, via the offer or answer.
+          </t>
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: Can you move this note up. The rest would be clearer if you understood that it only impacts new O/A transactions.

File: draft-ietf-rtcweb-jsep.xml [hunk 25]
```diff
@@ -1901,8 +1970,9 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
               <t>A direction attribute which is the same as that of the
               associated transceiver.</t>
 
-              <t>For each supported codec, "a=rtpmap" and "a=fmtp"
-              lines, as specified in
+              <t>For each supported codec that has not been excluded by
+              the codec preferences of the associated transceiver,
+              "a=rtpmap" and "a=fmtp" lines, as specified in
```
Comment (ekr) [inline draft-ietf-rtcweb-jsep.xml]: I think you would be better if you phrased this as
"For each codec that is available for this transceiver" rather than "supported filtered by excluded"

File: draft-ietf-rtcweb-jsep.xml (hunk 23 above)
Comment (taylor-b) [inline draft-ietf-rtcweb-jsep.xml]: Changed to "in turn affect" to make it unambiguous that this means "the preferences affect".

File: draft-ietf-rtcweb-jsep.xml (hunk 24 above)
Comment (taylor-b) [inline draft-ietf-rtcweb-jsep.xml]: Done. Also added something clarifying that excluding a codec doesn't mean you immediately stop sending it.

File: draft-ietf-rtcweb-jsep.xml (hunk 25 above)
Comment (taylor-b) [inline draft-ietf-rtcweb-jsep.xml]: As discussed I think this would require some restructuring (to define "codec available for a transceiver" somewhere) so we can save this for later, in the interest of landing the PR.

File: draft-ietf-rtcweb-jsep.xml [hunk 26]
```diff
@@ -1861,6 +1930,10 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             specified in
             <xref target="I-D.nandakumar-mmusic-proto-iana-registration" /> if
             the default candidate uses TCP transport.</t>
+
+            <t>If codec preferences have been set for the associated
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: I think you need to talk about excluded codecs here too.

File: draft-ietf-rtcweb-jsep.xml (hunk 25 above)
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: You might just want to say "for each media format specified on the m= line", since that should cover it.

File: draft-ietf-rtcweb-jsep.xml [hunk 27]
```diff
@@ -2501,6 +2579,10 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <t>The &lt;proto&gt; field MUST be set to exactly match the
             &lt;proto&gt; field for the corresponding m= line in the
             offer.</t>
+
+            <t>If codec preferences have been set for the associated
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: See note above on excluded codecs

File: draft-ietf-rtcweb-jsep.xml [hunk 28]
```diff
@@ -2531,7 +2613,9 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
             <t>A direction attribute which is the same as that of the
             associated transceiver.</t>
 
-            <t>For each supported codec that is present in the offer,
+            <t>For each supported codec that is present in the offer and
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: See note above about referencing m= line formats

File: draft-ietf-rtcweb-jsep.xml (hunk 26 above)
Comment (taylor-b) [inline draft-ietf-rtcweb-jsep.xml]: Done. Also moved "must support these mandatory codecs" up here.

File: draft-ietf-rtcweb-jsep.xml [hunk 29]
```diff
@@ -1974,15 +1983,9 @@ candidate:1 1 UDP 1694498815 192.0.2.33 10000 typ host
...
-              the codec preferences of the associated transceiver,
+              <t>For each media format on the m= line,
               "a=rtpmap" and "a=fmtp" lines, as specified in
-              <xref target="RFC4566"></xref>, Section 6. The audio and
-              video codecs that MUST be supported are specified in
-              <xref target="I-D.ietf-rtcweb-audio"></xref>(see Section
-              3) and
-              <xref target="I-D.ietf-rtcweb-video"></xref>(see Section
-              5).</t>
+              <xref target="RFC4566"></xref>, Section 6.</t>
```
Comment (juberti) [inline draft-ietf-rtcweb-jsep.xml]: suggest mentioning 3264, S 5.1, similar to the answer section below.
